Sources/App/Dem/DownloadDem.swift

```diff
@@ -71,96 +71,110 @@ struct Dem90 {
-                ]
-                do {
-                    try Process.spawnOrDie(cmd: "curl", args: args)
-                } catch {
-                    try "".write(toFile: testFile, atomically: true, encoding: .utf8)
-                    continue
-                }*/
-                
-                try Process.spawn(cmd: "gdal_translate", args: ["-of","NetCDF",tifLocal,ncTemp])
-                //try FileManager.default.removeItem(atPath: tifTemp)
-                
-                let data = try readNc(file: ncTemp)
-                try FileManager.default.removeItem(atPath: ncTemp)
-                
-
-                try OmFileWriter(dim0: data.dimensions[0], dim1: data.dimensions[1], chunk0: 20, chunk1: 20).write(file: omFile, compressionType: .p4nzdec256, scalefactor: 1, all: data.data)
             }
-        }
-        
-        for lat in -90..<90 {
-            if FileManager.default.fileExists(atPath: "\(Dem90.omDirectory)lat_\(lat).om") {
-                continue
-            }
-            let px = Dem90.pixel(latitude: lat)
-            var line = [Float](repeating: 0, count: 360*1200*px)
-            for lon in -180..<180 {
-                logger.info("Dem convert lon \(lon) lat \(lat)")
-                let omFile = "\(Dem90.downloadDirectory)\(lat)_\(lon).om"
-                if !FileManager.default.fileExists(atPath: omFile) {
-                    continue
+
+            try await group.waitForAll()
+
+            for lat in -90..<90 {
+                if FileManager.default.fileExists(atPath: "\(Dem90.omDirectory)lat_\(lat).om") {
+                    return
                 }
-                let om = try OmFileReader(file: omFile)
-                precondition(om.dim0 == 1200)
-                precondition(om.dim1 == px)
-                let data = try om.readAll()
-                for i in 0..<1200 {
-                    line[i * (px * 360) + (lon+180)*px ..< i * (px * 360) + (lon+180)*px + px] = data[i*px ..< (i+1)*px]
+
+                group.addTask {
+                    let px = Dem90.pixel(latitude: lat)
+                    var line = [Float](repeating: 0, count: 360*1200*px)
+                    for lon in -180..<180 {
+                        logger.info("Dem convert lon \(lon) lat \(lat)")
+                        let omFile = "\(Dem90.downloadDirectory)\(lat)_\(lon).om"
+                        if !FileManager.default.fileExists(atPath: omFile) {
+                            continue
+                        }
+
+                        let om = try OmFileReader(file: omFile)
+                        precondition(om.dim0 == 1200)
+                        precondition(om.dim1 == px)
+                        let data = try om.readAll()
+                        for i in 0..<1200 {
+                            line[i * (px * 360) + (lon+180)*px ..< i * (px * 360) + (lon+180)*px + px] = data[i*px ..< (i+1)*px]
+                        }
```

Comment:
So this actually exhausts memory when running concurrently, my understanding is that it's trying to mmap all the source files. Is there a way to rewrite this to use `om.read(into:)` so that it directly writes a subset of the data into `line`, without needing to `readAll`?

Reply:
Thanks for contributing! ‘om.read(into:)’ wont help you much as each job still needs its own buffers.

You could try to use a semaphore and limit the number of parallel jobs. Ideally the number of parallel jobs can then be specified with a CLI parameter (-c, —concurrent <n_jobs>). I think a default of 4 jobs should be reasonable. Wdyt?

Reply:
A semaphore and CLI flag makes sense to me - I can take a look, though if you have any pointers on which semaphore to use, let me know.

On buffers, that makes sense, though does the loop actually need _all_ of the data in each omFile in memory at the same time? My hope for not using `readAll` would be that the loops could read the data they need out of the mmap'd files rather than also buffering them into data as `[Float]`.

Reply:
I believe it needs all data. In theory it could be chunked, but this would require quite a rewrite.

You could try a simple to solution to limit the number of jobs: https://stackoverflow.com/a/70976324

Reply:
Okay, I think I've added two flags, `--concurrent-conversion-jobs` and `--concurrent-compression-jobs`, both of which default to 4.

In my case, the conversion jobs that call `gdal_translate` aren't memory-constrained, and I'm running this on a machine with 16 CPUs, hence the two flags rather than just a single one. I'd be able to run this with `open-meteo download-dem --concurrent-conversion-jobs 16` to raise the limit on just that phase.